core/Network/TLS/Types.hs

```diff
@@ -76,3 +79,27 @@ data Direction = Tx | Rx
 invertRole :: Role -> Role
 invertRole ClientRole = ServerRole
 invertRole ServerRole = ClientRole
+
+data SecretTriple = SecretTriple {
+    triBase   :: Secret13
+  , triClient :: TrafficSecret
+  , triServer :: TrafficSecret
+  } deriving (Eq, Show)
+
+data Secret13 = NoSecret
+              | EarlySecret ByteString
+              | HandshakeSecret ByteString
+              | ApplicationSecret ByteString -- TLS 1.3 master secret
+              | ResumptionSecret ByteString
+              deriving (Eq, Show)
+
+data TrafficSecret =
+    -- TLS 1.2 or earlier
+    MasterSecret12 ByteString
+    -- TLS 1.3
+  | ClientEarlySecret        ByteString
+  | ServerHandshakeSecret    ByteString
+  | ClientHandshakeSecret    ByteString
+  | ServerApplicationSecret0 ByteString
+  | ClientApplicationSecret0 ByteString
```

Comment:
This may be a problem when re-keying. Section 7.2 of rfc 8446 specify that the number will increase by one every time someone rekey. I would put the number as a parameter for the constructor.

Reply:
Good point. I will see if this is necessary.

Reply:
For logging it doesn't seem necessary.